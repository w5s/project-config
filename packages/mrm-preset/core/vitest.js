const fs = require('node:fs');
const { file, packageJson } = require('mrm-core');
const pkg = require('./pkg.js');
const npm = require('./npm.js');
const project = require('./project.js');

/**
 * @param {{
 *   state: 'present'|'absent'
 * }} options
 */
function vitest({ state }) {
  const hasVitest = state === 'present';
  const hasWorkspace = pkg.hasWorkspaces(packageJson());
  const vitestScript = 'vitest run';
  const update = (/** @type {string|undefined} */ _) =>
    hasVitest ? vitestScript : _ === vitestScript ? pkg.emptyScript : _;

  // Dependencies
  npm.dependency({
    dev: true,
    name: ['vite', 'vitest'],
    state: hasVitest ? 'present' : 'absent',
  });

  /** @type {string|undefined} */
  let viteVersion;
  /** @type {string|undefined} */
  let vitestVersion;
  pkg.withPackageJson((packageFile) => {
    viteVersion = packageFile.get('devDependencies.vite');
    vitestVersion = packageFile.get('devDependencies.vitest');
    pkg.script(packageFile, {
      name: `${project.test}:root`,
      update,
      state: hasWorkspace ? 'absent' : 'present',
    });
  });

  pkg.forEachWorkspace(({ projectDir, packageFile }) => {
    const hasSrc = fs.existsSync(`${projectDir}/src`);
    const packageState = hasVitest && hasSrc ? 'present' : 'absent';
    pkg.value(packageFile, {
      path: 'devDependencies.vite',
      state: packageState,
      update: (_) => _ ?? viteVersion,
    });
    pkg.value(packageFile, {
      path: 'devDependencies.vitest',
      state: packageState,
      update: (_) => _ ?? vitestVersion,
    });
    pkg.script(packageFile, {
      name: `${project.test}:src`,
      update,
      state: packageState,
    });
    const viteConfig = file(`${projectDir}/vite.config.mts`);
    if (packageState === 'absent') {
      viteConfig.delete();
    } else if (!viteConfig.exists()) {
      viteConfig.save(
        `/// <reference types="vitest" />
import { defineConfig } from 'vite';

export default defineConfig({
  test: {},
});
`
      );
    }
  });
}

module.exports = {
  vitest,
};
