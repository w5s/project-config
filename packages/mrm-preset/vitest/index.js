const { vitest } = require('../core/vitest.js');

/**
 *
 */
function task() {
  vitest({
    state: 'present',
  });
}

task.description = 'Setup Vitest';
task.parameters = {};

module.exports = task;
