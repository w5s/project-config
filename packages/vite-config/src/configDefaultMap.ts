import { UserConfigExport, configDefaults } from 'vitest/config';
import { ConfigType } from './type.js';

const defaultConfig = {
  test: configDefaults,
};

export const configDefaultMap: { readonly [T in ConfigType]: UserConfigExport } = Object.freeze({
  application: defaultConfig,
  library: defaultConfig,
});
