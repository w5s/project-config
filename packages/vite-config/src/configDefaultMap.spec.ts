import { describe, it, expect } from 'vitest';
import { configDefaultMap } from './configDefaultMap.js';

describe('configDefaultMap', () => {
  it('should be an object', () => {
    expect(configDefaultMap).toMatchSnapshot();
  });
});
