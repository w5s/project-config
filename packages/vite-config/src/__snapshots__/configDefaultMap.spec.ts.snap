// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`configDefaultMap > should be an object 1`] = `
{
  "application": {
    "test": {
      "allowOnly": true,
      "api": false,
      "clearMocks": false,
      "coverage": {
        "clean": true,
        "cleanOnRerun": true,
        "enabled": false,
        "exclude": [
          "coverage/**",
          "dist/**",
          "packages/*/test?(s)/**",
          "**/*.d.ts",
          "cypress/**",
          "test?(s)/**",
          "test?(-*).?(c|m)[jt]s?(x)",
          "**/*{.,-}{test,spec}.?(c|m)[jt]s?(x)",
          "**/__tests__/**",
          "**/{karma,rollup,webpack,vite,vitest,jest,ava,babel,nyc,cypress,tsup,build}.config.*",
          "**/.{eslint,mocha,prettier}rc.{?(c|m)js,yml}",
        ],
        "extension": [
          ".js",
          ".cjs",
          ".mjs",
          ".ts",
          ".mts",
          ".cts",
          ".tsx",
          ".jsx",
          ".vue",
          ".svelte",
        ],
        "provider": "v8",
        "reportOnFailure": true,
        "reporter": [
          [
            "text",
            {},
          ],
          [
            "html",
            {},
          ],
          [
            "clover",
            {},
          ],
          [
            "json",
            {},
          ],
        ],
        "reportsDirectory": "./coverage",
      },
      "css": {
        "include": [],
      },
      "dangerouslyIgnoreUnhandledErrors": false,
      "environment": "node",
      "exclude": [
        "**/node_modules/**",
        "**/dist/**",
        "**/cypress/**",
        "**/.{idea,git,cache,output,temp}/**",
        "**/{karma,rollup,webpack,vite,vitest,jest,ava,babel,nyc,cypress,tsup,build}.config.*",
      ],
      "fakeTimers": {
        "loopLimit": 10000,
        "shouldClearNativeTimers": true,
        "toFake": [
          "setTimeout",
          "clearTimeout",
          "setInterval",
          "clearInterval",
          "setImmediate",
          "clearImmediate",
          "Date",
        ],
      },
      "forceRerunTriggers": [
        "**/package.json/**",
        "**/{vitest,vite}.config.*/**",
      ],
      "globals": false,
      "hideSkippedTests": false,
      "hookTimeout": 10000,
      "include": [
        "**/__tests__/**/*.?(c|m)[jt]s?(x)",
        "**/?(*.){test,spec}.?(c|m)[jt]s?(x)",
      ],
      "isolate": true,
      "maxConcurrency": 5,
      "mockReset": false,
      "open": true,
      "reporters": [],
      "restoreMocks": false,
      "silent": false,
      "slowTestThreshold": 300,
      "teardownTimeout": 10000,
      "testTimeout": 5000,
      "threads": true,
      "typecheck": {
        "checker": "tsc",
        "exclude": [
          "**/node_modules/**",
          "**/dist/**",
          "**/cypress/**",
          "**/.{idea,git,cache,output,temp}/**",
          "**/{karma,rollup,webpack,vite,vitest,jest,ava,babel,nyc,cypress,tsup,build}.config.*",
        ],
        "include": [
          "**/?(*.){test,spec}-d.?(c|m)[jt]s?(x)",
        ],
      },
      "ui": false,
      "uiBase": "/__vitest__/",
      "update": false,
      "watch": true,
      "watchExclude": [
        "**/node_modules/**",
        "**/dist/**",
      ],
    },
  },
  "library": {
    "test": {
      "allowOnly": true,
      "api": false,
      "clearMocks": false,
      "coverage": {
        "clean": true,
        "cleanOnRerun": true,
        "enabled": false,
        "exclude": [
          "coverage/**",
          "dist/**",
          "packages/*/test?(s)/**",
          "**/*.d.ts",
          "cypress/**",
          "test?(s)/**",
          "test?(-*).?(c|m)[jt]s?(x)",
          "**/*{.,-}{test,spec}.?(c|m)[jt]s?(x)",
          "**/__tests__/**",
          "**/{karma,rollup,webpack,vite,vitest,jest,ava,babel,nyc,cypress,tsup,build}.config.*",
          "**/.{eslint,mocha,prettier}rc.{?(c|m)js,yml}",
        ],
        "extension": [
          ".js",
          ".cjs",
          ".mjs",
          ".ts",
          ".mts",
          ".cts",
          ".tsx",
          ".jsx",
          ".vue",
          ".svelte",
        ],
        "provider": "v8",
        "reportOnFailure": true,
        "reporter": [
          [
            "text",
            {},
          ],
          [
            "html",
            {},
          ],
          [
            "clover",
            {},
          ],
          [
            "json",
            {},
          ],
        ],
        "reportsDirectory": "./coverage",
      },
      "css": {
        "include": [],
      },
      "dangerouslyIgnoreUnhandledErrors": false,
      "environment": "node",
      "exclude": [
        "**/node_modules/**",
        "**/dist/**",
        "**/cypress/**",
        "**/.{idea,git,cache,output,temp}/**",
        "**/{karma,rollup,webpack,vite,vitest,jest,ava,babel,nyc,cypress,tsup,build}.config.*",
      ],
      "fakeTimers": {
        "loopLimit": 10000,
        "shouldClearNativeTimers": true,
        "toFake": [
          "setTimeout",
          "clearTimeout",
          "setInterval",
          "clearInterval",
          "setImmediate",
          "clearImmediate",
          "Date",
        ],
      },
      "forceRerunTriggers": [
        "**/package.json/**",
        "**/{vitest,vite}.config.*/**",
      ],
      "globals": false,
      "hideSkippedTests": false,
      "hookTimeout": 10000,
      "include": [
        "**/__tests__/**/*.?(c|m)[jt]s?(x)",
        "**/?(*.){test,spec}.?(c|m)[jt]s?(x)",
      ],
      "isolate": true,
      "maxConcurrency": 5,
      "mockReset": false,
      "open": true,
      "reporters": [],
      "restoreMocks": false,
      "silent": false,
      "slowTestThreshold": 300,
      "teardownTimeout": 10000,
      "testTimeout": 5000,
      "threads": true,
      "typecheck": {
        "checker": "tsc",
        "exclude": [
          "**/node_modules/**",
          "**/dist/**",
          "**/cypress/**",
          "**/.{idea,git,cache,output,temp}/**",
          "**/{karma,rollup,webpack,vite,vitest,jest,ava,babel,nyc,cypress,tsup,build}.config.*",
        ],
        "include": [
          "**/?(*.){test,spec}-d.?(c|m)[jt]s?(x)",
        ],
      },
      "ui": false,
      "uiBase": "/__vitest__/",
      "update": false,
      "watch": true,
      "watchExclude": [
        "**/node_modules/**",
        "**/dist/**",
      ],
    },
  },
}
`;
